cloud-controller-manager/do/resources.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2017 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	v1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	v1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+)
+
+const (
+	controllerSyncPeriod     = 1 * time.Minute
+	requestTimeout           = 10 * time.Second
+	resourceTypeLoadBalancer = "load_balancer"
+)
+
+// ResourcesController ensures that DO resources are properly tagged.
+type ResourcesController struct {
+	clusterIDTag string
+	kclient      kubernetes.Interface
+	gclient      *godo.Client
+	svcLister    v1lister.ServiceLister
+}
+
+// NewResourcesController returns a new services controller responsible for
+// re-applying cluster IDs on DO load-balancers.
+func NewResourcesController(clusterIDTag string, inf v1informers.ServiceInformer,
+	k kubernetes.Interface, g *godo.Client) *ResourcesController {
+	return &ResourcesController{
+		clusterIDTag: clusterIDTag,
+		kclient:      k,
+		gclient:      g,
+		svcLister:    inf.Lister(),
+	}
+}
+
+// Run starts the resources controller. It watches over DigitalOcean resources
+// making sure the right tags are set.
+func (r *ResourcesController) Run(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(controllerSyncPeriod)
+	defer ticker.Stop()
+	// Do not wait for initial tick to pass; run immediately for reduced sync
+	// latency.
+	for tickerC := ticker.C; ; {
+		if err := r.sync(); err != nil {
+			glog.Errorf("failed to sync load-balancer tags: %s", err)
+		}
+		select {
+		case <-tickerC:
+			continue
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (r *ResourcesController) sync() error {
+	// Fetch all load-balancers.
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	lbs, _, err := r.gclient.LoadBalancers.List(ctx, &godo.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list load-balancers: %s", err)
+	}
+
+	// Collect tag resources for known load balancer names (i.e., services
+	// with type=LoadBalancer.)
+	svcs, err := r.svcLister.List(labels.Everything())
+	if err != err {
+		return fmt.Errorf("failed to list services: %s", err)
+	}
+	var res []godo.Resource
+	for _, lb := range lbs {
+		for _, svc := range svcs {
```

Comment:
This loop has some nice quadratic performance, you should use a map here to make it O(N) instead of O(MN)

Reply:
I think it'd still be O(MN) even if we kept track of visited services as we go through the elements. (Practically, it will be faster because the inner loop needs to run through all services only once per outer loop.)

In order to make this truly O(N), I believe we'd have to manage a map that is modified as service objects join and leave in the cluster. However, that'd require extra synchronization and more testing.

I'm not sure if it's worth the effort though given that practically speaking, quadratic time probably won't bother us before a few other bottlenecks. WDYT?

Reply:
https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/185